### ISSUE TITLE
Negative index numbers in case of /64 IPv6 subnet

### ISSUE THREAD

**Issue by TothFerenc** [k8snetworkplumbingwg/whereabouts#83]
In case we use quite big IPv6 subnets (like /64) during the IP allocation the calculated index value can overflow to a negative number.
Example NAD:
```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
 name: ipvlan-whereabouts
 namespace: kube-system
spec:
 config: '{
  "cniVersion": "0.3.1",
  "type": "ipvlan",
  "master": "ens1f0",
  "mode": "l2",
  "ipam": {
   "type": "whereabouts",
   "range": "caa5::0/64",
   "range_start": "caa5:0:0:0:caa5::0",
   "range_end": "caa5:0:0:0:caa5::10"
  }
 }'
```
IPPool content:
```
# kubectl get ippool caa5---64 -n kube-system -o yaml
apiVersion: whereabouts.cni.cncf.io/v1alpha1
kind: IPPool
metadata:
[..]
  name: caa5---64
  namespace: kube-system
  resourceVersion: "8139228"
  selfLink: /apis/whereabouts.cni.cncf.io/v1alpha1/namespaces/kube-system/ippools/caa5---64
  uid: 469d8ea7-60fa-4317-82c1-6d14b0e41221
spec:
  allocations:
    "-3844666706890850303":
      id: 5aa4be0cb5625dd9912228ce232f23aa9b1870fdcfe60c714325f4b407d09d28
  range: caa5::/64
```
Is it possible to use uint64 to store index value? Maybe [uint128](https://github.com/lukechampine/uint128) would be even better.
Thanks!

**Comment by dougbtv** [comment]
@TothFerenc thanks for filing the issue -- and for your keen eye on Whereabouts issues as a whole, I appreciate it!

Good call, I think uint128 probably would work great.

**Comment by dougbtv** [comment]
@TothFerenc -- do you mind checking out if #89 fixes this up for you?

We refactored how we calculate IPv6 addresses in this one, and would love to know if this logic hit your case. Thanks!

**Comment by TothFerenc** [comment]
Yes. The reported case looks better now:
```
    allocations:
      "14602077366818701312":
        id: a8c9471e301ce26ab909bdf3dd8840b77e21b09b7128d8ccff9c3aa2202610ac
    range: caa5::/64
```
With the new implementation how wide can be an IPv6 subnet?
Does the new arithmetic backward compatible (considering those cases when the old method worked)? Can we upgrade a system by replacing the whereabouts binary while keeping the IP allocations in IPPools?
Thanks!

**Comment by TothFerenc** [comment]
I got the same offset number for the below cases: IPs are different!
/64:
```
4: net1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8900 qdisc noqueue
    link/ether 00:50:56:b0:c0:e2 brd ff:ff:ff:ff:ff:ff
    inet6 caa5::caa5:0:0:0/64 scope global
[..]
    allocations:
      "14602077366818701312":
        id: a8c9471e301ce26ab909bdf3dd8840b77e21b09b7128d8ccff9c3aa2202610ac
    range: caa5::/64
```
/63:
```
4: net1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8900 qdisc noqueue
    link/ether 00:50:56:b0:2a:a3 brd ff:ff:ff:ff:ff:ff
    inet6 caa5:0:0:1:caa5::/63 scope global
[..]
    allocations:
      "14602077366818701312":
        id: 166118388d1b2c35385747b05f9a93ebc5b65f74615d331c715b7fd7f03c0afa
    range: caa5::/63
```

**Comment by TothFerenc** [comment]
@dougbtv
This is still a valid issue. In case of /64 subnet, only the first half of the range can be addressed by Whereabouts (v0.5):
from `x:x:x:x::0` to `x:x:x:x:7fff:ffff:ffff:ffff`
But I understand this is something we can live with it :)
Nevertheless it is worth to document it.
